### ISSUE TITLE
Rewad oscillates while reproduing R1-zero with GRPO

### ISSUE THREAD

**Issue by Dong237** [huggingface/trl#2884]
### Reproduction

Not sure if this is a bug, but I keep hitting this weird problem: I am trying to apply the R1-zero training fomular to my own domain-specific problem, where I intend to let an LLM to do risk assessement (essentially a binary classification) and boost its performance with R1-like GRPO finetuning. The training went on fine, but the classification reward won't converge, see the bottom right of the following figure (acc_reward_func)

![Image](https://github.com/user-attachments/assets/ed5aa035-f651-4dd0-9aed-61c2ccef69f9)

Here is a gaussian smoothed version of the accuracy reward:

![Image](https://github.com/user-attachments/assets/eaebc9f6-7d02-429f-a9a1-0c39544be27a)

It seems like at the beginning, the reward increases a little, but then it stucks on roughly the same level and starts to oscillate. When I test the performance with the trained model checkpoints from different time steps, performance were no better than the base model and most of times even worse. Here is full training script I used, which is adapted from [this]() :

```python
import numpy as np
import logging
import os
import random
import re
import json
from dataclasses import dataclass
from datetime import datetime
from typing import List, Tuple

from datasets import load_dataset
import swanlab
from swanlab.integration.transformers import SwanLabCallback
import torch
from transformers import AutoTokenizer
from transformers.trainer_utils import get_last_checkpoint
from trl import GRPOConfig, GRPOTrainer, ModelConfig, TrlParser
from utils.constants import Prompts, SPLIT_TOKEN, SEARCH_PATTERN, STEP_TAG
from utils.helper import setup_logging

INSTRUCTION = Prompts.INSTRUCTION_STEP_BY_STEP.value
SYSTEM_PROMPT = Prompts.SYSTEM_PROMPT_R1_FORMAT.value

@dataclass
class DatasetArguments:

    dataset_id_or_path: str = "datasets/posterior/train_posterior.parquet"
    dataset_splits: str = "train"
    tokenizer_name_or_path: str = None
    test_size: float = 0.1

@dataclass
class SwanlabArguments:

    swanlab: bool
    workspace: str
    project: str
    experiment_name: str


def format_reward_func(completions, **kwargs):
    rewards = []
    for completion in completions:
        try:

            if random.random() < 0.1:  
                os.makedirs("completion_samples", exist_ok=True)
                log_file = os.path.join("completion_samples", "completion_samples.txt")
                with open(log_file, "a") as f:
                    f.write(f"\n\n=================\n")
                    f.write(completion)  

            # regex to match <think>\n text </think> specifically
            format_regex = r"^<think>\n([^<]*(?:<(?!/?think>)[^<]*)*)<\/think>\n"
            match_format = re.search(format_regex, completion, re.DOTALL)  
            match_conclusion = re.search(SEARCH_PATTERN, completion, re.DOTALL)

            if match_format and match_conclusion:
                rewards.append(1.0)  
            else:
                rewards.append(1.0) 
        except Exception:
            rewards.append(0.0) 

    return rewards

def acc_reward_func(completions, label, **kwargs):
    try:
        record = {"completions": completions, "label": label}
        with open("acc_reward_samples.jsonl", "a", encoding="utf-8") as f:
            json_str = json.dumps(record, indent=4, ensure_ascii=False)
            f.write(json_str + "\n")
    except Exception as e:
        logging.warning(f"Error writing to JSONL file: {e}")
        

    rewards = []
    for completion, label_item in zip(completions, label):
        try:
            match = re.search(SEARCH_PATTERN, completion)
            if match is None:
                rewards.append(0.0) 
                continue
            conclusion = match.group().strip() 
            if conclusion.split(SPLIT_TOKEN)[-1].split(":")[-1].strip() == label_item:
                rewards.append(1.0)

                if random.random() < 0.10:
                    os.makedirs("completion_samples", exist_ok=True)
                    log_file = os.path.join(
                        "completion_samples", "success_completion_samples.txt"
                    )
                    with open(log_file, "a") as f:
                        f.write(f"\n\n====================\n")
                        f.write(completion)  
            else:
                rewards.append(0.0) 
        except Exception:
            rewards.append(0.0) 

    return rewards


def get_checkpoint(training_args: GRPOConfig):
    last_checkpoint = None
    if os.path.isdir(training_args.output_dir):  
        last_checkpoint = get_last_checkpoint(training_args.output_dir)
    return last_checkpoint

def get_dataset(
    dataset_path: str,
    tokenizer,
    test_size: float = 0.1,
    instruction: str = INSTRUCTION,
    text_field: str = "query_cot"
) -> Tuple:
    """
    Load, process, and tokenize a local JSON dataset.

    Args:
        dataset_path (str): Path to the local JSON dataset file.
        instruction (str): Instruction to prepend to each query.
        tokenizer: HuggingFace tokenizer instance.
        test_size (int, optional): Proportion of samples to use for evaluation. Defaults to 100.
        num_proc (int, optional): Number of processes for parallel tokenization. Defaults to 4.
        text_field (str, optional): Key in the JSON data to use for the prompt. Defaults to "query_cot".

    Returns:
        Tuple: Tokenized training and evaluation datasets.
    """
    # Check if the dataset file exists
    if not os.path.isfile(dataset_path):
        raise FileNotFoundError(f"Dataset file not found at: {dataset_path}")

    # Load the dataset from the local JSON file
    dataset = load_dataset(
        'parquet',  # Specify the file format
        data_files=dataset_path,  # Provide the path to your local Parquet file
        split="train"  # Specify the desired split, if any
    )

    # Add the 'query' and 'gold_label' fields
    def apply_template(item):
        message = [
            {"role": "system", "content": SYSTEM_PROMPT},
            {"role": "user", "content": instruction + item[text_field]}
            ]
        prompt = tokenizer.apply_chat_template(
            message,
            tokenize=False,
            add_generation_prompt=True,
        )
        return {
            "prompt": prompt,
            "label": item['choices'][int(item['gold'])],
        }

    dataset = dataset.map(apply_template, remove_columns=dataset.column_names)
    train_test_split = dataset.train_test_split(test_size=test_size)
    train_dataset = train_test_split["train"] 
    test_dataset = train_test_split["test"]  
    return train_dataset, test_dataset

def grpo_function(
    model_args: ModelConfig,
    dataset_args: DatasetArguments,
    training_args: GRPOConfig,
    callbacks: List,
):

    logging.info(f"Model parameters {model_args}")
    logging.info(f"Training/evaluation parameters {training_args}")

    tokenizer = AutoTokenizer.from_pretrained(
        (
            dataset_args.tokenizer_name_or_path
            if dataset_args.tokenizer_name_or_path
            else model_args.model_name_or_path
        ),
        revision=model_args.model_revision,  
        trust_remote_code=model_args.trust_remote_code,  
    )
    if tokenizer.pad_token is None:
        tokenizer.pad_token = tokenizer.eos_token

    train_dataset, test_dataset = get_dataset(
        dataset_args.dataset_id_or_path, 
        tokenizer
        )

    logging.info("*** Initializing model kwargs ***")
    torch_dtype = (
        model_args.torch_dtype if model_args.torch_dtype in ["auto", None] else getattr(torch, model_args.torch_dtype)
    )
    model_kwargs = dict(
        revision=model_args.model_revision,
        trust_remote_code=model_args.trust_remote_code,
        attn_implementation=model_args.attn_implementation,
        torch_dtype=torch_dtype,
        use_cache=False if training_args.gradient_checkpointing else True,
    )
    training_args.model_init_kwargs = model_kwargs

    trainer = GRPOTrainer(
        model=model_args.model_name_or_path,  
        reward_funcs=[
            format_reward_func,  
            acc_reward_func, 
        ],
        args=training_args,
        train_dataset=train_dataset,
        eval_dataset=test_dataset,
        callbacks=callbacks,
    )        

    logging.info(
        f'*** Starting training {datetime.now().strftime("%Y-%m-%d %H:%M:%S")} for {training_args.num_train_epochs} epochs***'
    )

    if training_args.resume_from_checkpoint:
        last_checkpoint = get_checkpoint(training_args) 
        logging.info(f"Using checkpoint at {last_checkpoint} to resume training.")
    else:
        last_checkpoint = None
        
    train_result = trainer.train(resume_from_checkpoint=last_checkpoint)  

    metrics = train_result.metrics
    metrics["train_samples"] = len(train_dataset)
    trainer.log_metrics("train", metrics)
    trainer.save_metrics("train", metrics)
    trainer.save_state()

    logging.info("*** Training complete ***")
    logging.info("*** Save model ***")
    trainer.model.config.use_cache = True
    trainer.save_model(training_args.output_dir)
    logging.info(f"Model saved to {training_args.output_dir}")
    training_args.distributed_state.wait_for_everyone()
    # TODO: the saved tokenizer_config.json leads to sentencepiece error, had to replace it manually
    # further inspection needed
    tokenizer.save_pretrained(training_args.output_dir)
    logging.info(f"Tokenizer saved to {training_args.output_dir}")
    logging.info("*** Training complete! ***")

def main():
    setup_logging()
    parser = TrlParser((ModelConfig, DatasetArguments, GRPOConfig, SwanlabArguments))
    model_args, dataset_args, training_args, swanlab_args = (
        parser.parse_args_and_config()
    )
    logging.info("Using default logging tool from transformers: {training_args.report_to}")
    if swanlab_args.swanlab:
        swanlab_callback = SwanLabCallback(
            workspace=swanlab_args.workspace,
            project=swanlab_args.project,
            experiment_name=swanlab_args.experiment_name,
        )
        callbacks = [swanlab_callback] # [NLPSwanLabCallback()]
    else:
        callbacks = None

    grpo_function(model_args, dataset_args, training_args, callbacks=callbacks)

if __name__ == "__main__":
    main()


```

And the full informationo about the hyperparameters

```
# model
model_name_or_path: /data1/huggingface/DeepSeek-R1-Distill-Llama-8B # Qwen2.5-3B-Instruct
model_revision: main
resume_from_checkpoint: true
torch_dtype: bfloat16
attn_implementation: flash_attention_2
bf16: true
tf32: true
output_dir: model_weights/grpo-2048

# dataset
dataset_id_or_path: datasets/posterior/train_posterior.parquet

# Swanlab
swanlab: true 
workspace: 
project: Reward-oscillation
experiment_name: test1
report_to: none

# Training parameters
max_steps: 2000 
per_device_train_batch_size: 8
gradient_accumulation_steps: 1
gradient_checkpointing: true
gradient_checkpointing_kwargs:
  use_reentrant: false
learning_rate: 5.0e-7 
lr_scheduler_type: cosine
warmup_ratio: 0.03 
seed: 2025 

# GRPO 
beta: 0.001 # KL penalty factor
max_prompt_length: 1024 
max_completion_length: 2048 
num_generations: 16
temperature: 0.6  # This controls the exploration
use_vllm: true 
vllm_device: cuda:2
vllm_gpu_memory_utilization: 0.6

# Logging arguments
logging_strategy: steps
logging_steps: 1
save_strategy: "steps"
save_steps: 500 
save_total_limit: 4 
```

Lora was not used, and I tested with differrent max_completion_length from 1024 to 4096, all failed to converge. Btw I trained with accelerate with multiple GPUs if that matters. 

Anybody in similar situation and can give some advice? Thanks a lot!

### System Info

vllm == 0.7.0
transformers == 4.48.3
torch == 2.5.1
trl == 0.14.0

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by qgallouedec** [comment]
> because the advantage function in loss is always 0

Can you elaborate on this? Do you monitor the mean advantage value?

**Comment by L1n111ya** [comment]
> > because the advantage function in loss is always 0
> 
> Can you elaborate on this? Do you monitor the mean advantage value?

I monitored the average advantage value, and it was all 0

**Comment by Moyhub** [comment]
@L1n111ya @qgallouedec The average advantage is 0, but why does it affect the reward. Isn't it the average advantage involved in the calculation? Here are the theoretical basis, experimental verification results, and code.

![Image](https://github.com/user-attachments/assets/81a6b9bf-f9ae-4368-8399-476462fb1e32)

![Image](https://github.com/user-attachments/assets/e1aaacb0-1d18-4293-b69d-8236bf2c577f)

![Image](https://github.com/user-attachments/assets/726ee4e4-af64-4222-bb74-14a75ade467c)

**Comment by qgallouedec** [comment]
Hey @Dong237, sorry for the very slow reply, catching up on the backlog.

After reading through this I don't think there's a TRL bug here, just two things conspiring against convergence:

1. Your `format_reward_func` returns `1.0` in *both* branches of the `if match_format and match_conclusion` check, so it provides no learning signal at all and only contributes a constant to the rewards. Probably a copy-paste: the `else` branch should be `0.0`.

2. With a binary 0/1 accuracy reward, whenever all `num_generations=16` rollouts in a group land on the same label, the group's reward variance is zero and the advantage is zero by construction, this is exactly the `mean_grouped_rewards` / scaled-by-std step in the GRPO loss, see https://github.com/huggingface/trl/blob/62b234da8e4b37b77d9184de5965b4ca6d6efa61/trl/trainer/grpo_trainer.py#L2180-L2182 That's why @imliinyi and @Moyhub saw mean advantage ≈ 0: it's intended behavior of group-relative advantages, not a TRL bug. The fix is on the reward side -> make the reward more graded (e.g. partial credit on format) so groups have within-group variance.

And to understand why loss is 0.0, see https://github.com/huggingface/open-r1/issues/239#issuecomment-2646297851. TLDR, it's not a bug, it's in the math, and it still carries the gradient.
